server/src/main/java/org/opensearch/common/cache/store/EhCacheDiskCache.java

```diff
@@ -0,0 +1,543 @@
+        this.expireAfterAccess = Objects.requireNonNull(builder.getExpireAfterAcess(), "ExpireAfterAccess value shouldn't " + "be null");
+        this.maxWeightInBytes = builder.getMaxWeightInBytes();
+        if (this.maxWeightInBytes <= MINIMUM_MAX_SIZE_IN_BYTES) {
+            throw new IllegalArgumentException("Ehcache Disk tier cache size should be greater than " + MINIMUM_MAX_SIZE_IN_BYTES);
+        }
+        this.storagePath = Objects.requireNonNull(builder.storagePath, "Storage path shouldn't be null");
+        if (builder.threadPoolAlias == null || builder.threadPoolAlias.isBlank()) {
+            this.threadPoolAlias = THREAD_POOL_ALIAS_PREFIX + "DiskWrite";
+        } else {
+            this.threadPoolAlias = builder.threadPoolAlias;
+        }
+        this.settings = Objects.requireNonNull(builder.settings, "Settings objects shouldn't be null");
+        Objects.requireNonNull(builder.settingPrefix, "Setting prefix shouldn't be null");
+        this.DISK_WRITE_MINIMUM_THREADS = Setting.intSetting(builder.settingPrefix + ".tier.disk.ehcache.min_threads", 2, 1, 5);
+        this.DISK_WRITE_MAXIMUM_THREADS = Setting.intSetting(builder.settingPrefix + ".tier.disk.ehcache.max_threads", 2, 1, 20);
+        // Default value is 1 within EhCache.
+        this.DISK_WRITE_CONCURRENCY = Setting.intSetting(builder.settingPrefix + ".tier.disk.ehcache.concurrency", 2, 1, 3);
+        // Default value is 16 within Ehcache.
+        this.DISK_SEGMENTS = Setting.intSetting(builder.settingPrefix + "tier.disk.ehcache.segments", 16, 1, 32);
+        this.cacheManager = buildCacheManager();
+        Objects.requireNonNull(builder.getEventListener(), "Listener can't be null");
+        this.eventListener = builder.getEventListener();
+        this.ehCacheEventListener = new EhCacheEventListener<K, V>(builder.getEventListener());
+        this.cache = buildCache(Duration.ofMillis(expireAfterAccess.getMillis()), builder);
+    }
+
+    @SuppressForbidden(reason = "Ehcache uses File.io")
+    private PersistentCacheManager buildCacheManager() {
+        // In case we use multiple ehCaches, we can define this cache manager at a global level.
+        return CacheManagerBuilder.newCacheManagerBuilder()
+            .with(CacheManagerBuilder.persistence(new File(storagePath)))
+            .using(
+                PooledExecutionServiceConfigurationBuilder.newPooledExecutionServiceConfigurationBuilder()
+                    .defaultPool(THREAD_POOL_ALIAS_PREFIX + "Default", 1, 3) // Default pool used for other tasks like
+                    // event listeners
+                    .pool(this.threadPoolAlias, DISK_WRITE_MINIMUM_THREADS.get(settings), DISK_WRITE_MAXIMUM_THREADS.get(settings))
+                    .build()
+            )
+            .build(true);
+    }
+
+    private Cache<K, V> buildCache(Duration expireAfterAccess, Builder<K, V> builder) {
+        return this.cacheManager.createCache(
+            DISK_CACHE_ALIAS,
```

Comment:
What happens if the disk cache already exists?  Seems like it should be not reused likely deleted.

Reply:
Yeah it will throw an exception if cache with same alias name already exists. Handled that.